File: lib/python/opentoken-cli/src/main/opentoken_cli/commands/tokenize_command.py
"""
Copyright (c) Truveta. All rights reserved.
"""

import logging
from typing import List

from opentoken.metadata import Metadata
from opentoken.tokens.tokenizer.passthrough_tokenizer import PassthroughTokenizer
from opentoken.tokentransformer.hash_token_transformer import HashTokenTransformer
from opentoken.tokentransformer.token_transformer import TokenTransformer
from opentoken_cli.io.csv.person_attributes_csv_reader import PersonAttributesCSVReader
from opentoken_cli.io.csv.person_attributes_csv_writer import PersonAttributesCSVWriter
from opentoken_cli.io.json.metadata_json_writer import MetadataJsonWriter
from opentoken_cli.io.parquet.person_attributes_parquet_reader import (
    PersonAttributesParquetReader,
)
from opentoken_cli.io.parquet.person_attributes_parquet_writer import (
    PersonAttributesParquetWriter,
)
from opentoken_cli.processor.person_attributes_processor import (
    PersonAttributesProcessor,
)
from opentoken_cli.util import StringMaskingUtil

logger = logging.getLogger(__name__)


class TokenizeCommand:
    """
    Tokenize command - generates tokens from person attributes.

    Normal mode (default): applies SHA-256 then HMAC-SHA256 hashing on the token
    signature, producing opaque base64 tokens. ``--hashingsecret`` is required.

    Demo mode (``--demo-mode``): skips all hashing so tokens are the raw
    pipe-separated attribute signature strings. No secret is needed, making it easy
    to explore the output without managing secrets. Demo-mode output is
    **not** suitable for production or cross-organisation exchange.
    """

    TYPE_CSV = "csv"
    TYPE_PARQUET = "parquet"

    @staticmethod
    def register_subcommand(subparsers):
        """Register the tokenize subcommand with the argument parser."""
        parser = subparsers.add_parser(
            "tokenize",
            help="Generate tokens from person attributes (normal mode: HMAC-SHA256; demo mode: plain signatures)",
            description=(
                "Generate tokens from person attributes.\n\n"
                "Normal mode: tokens are HMAC-SHA256 hashed (--hashingsecret required).\n"
                "Demo mode (--demo-mode): tokens are plain attribute signature strings; no secret needed."
            ),
            add_help=False,  # Disable automatic -h for help to allow -h for hashingsecret
        )

        # Manually add --help (without -h short form)
        parser.add_argument(
            "--help",
            action="help",
            help="Show this help message and exit",
        )

        parser.add_argument(
            "-i",
            "--input",
            required=True,
            dest="input_path",
            help="Input file path",
        )

        parser.add_argument(
            "-o",
            "--output",
            required=True,
            dest="output_path",
            help="Output file path",
        )

        parser.add_argument(
            "-t",
            "--input-type",
            required=True,
            dest="input_type",
            choices=["csv", "parquet"],
            help="Input file type: csv or parquet",
        )

        parser.add_argument(
            "-ot",
            "--output-type",
            dest="output_type",
            choices=["csv", "parquet"],
            help="Output file type (defaults to input type): csv or parquet",
        )

        parser.add_argument(
            "--demo-mode",
            action="store_true",
            default=False,
            dest="demo_mode",
            help=(
                "Enable demo mode: output plain attribute signature strings without any hashing. "
                "--hashingsecret is not required in this mode. "
                "Demo output is NOT suitable for production or cross-organisation exchange."
            ),
        )

        # Optional in demo mode; required at runtime in normal mode
        parser.add_argument(
            "-h",
            "--hashingsecret",
            required=False,
            dest="hashing_secret",
            help="Hashing secret for HMAC-SHA256 token generation (required in normal mode, ignored in demo mode)",
        )

        parser.set_defaults(func=TokenizeCommand.execute)

    @staticmethod
    def execute(args):
        """Execute the tokenize command."""
        demo_mode = getattr(args, "demo_mode", False)

        if demo_mode:
            logger.warning(
                "Running in DEMO MODE - tokens are plain attribute signature strings, not HMAC-hashed. "
                "Do not use demo-mode output in production or share it externally."
            )
        else:
            logger.info("Running tokenize command (normal mode)")

        # Default output type to input type if not specified
        output_type = args.output_type if args.output_type else args.input_type

        logger.info(f"Input: {args.input_path} ({args.input_type})")
        logger.info(f"Output: {args.output_path} ({output_type})")
        if not demo_mode:
            logger.info(
                f"Hashing Secret: {StringMaskingUtil.mask_string(args.hashing_secret)}"
            )

        # --hashingsecret is required in normal mode only
        if not demo_mode:
            if not args.hashing_secret or not args.hashing_secret.strip():
                logger.error(
                    "--hashingsecret is required in normal mode. Use --demo-mode to skip hashing."
                )
                return 1

        try:
            if demo_mode:
                TokenizeCommand._process_tokens_demo(
                    args.input_path,
                    args.output_path,
                    args.input_type,
                    output_type,
                )
            else:
                TokenizeCommand._process_tokens(
                    args.input_path,
                    args.output_path,
                    args.input_type,
                    output_type,
                    args.hashing_secret,
                )
            logger.info("Token generation completed successfully")
            return 0
        except Exception as e:
            logger.error(f"Error during token generation: {e}", exc_info=True)
            return 1

    @staticmethod
    def _process_tokens(
        input_path: str,
        output_path: str,
        input_type: str,
        output_type: str,
        hashing_secret: str,
    ):
        """Process tokens in normal mode using SHA-256 + HMAC-SHA256."""
        token_transformer_list: List[TokenTransformer] = []

        try:
            # Add only hash transformer (no encryption in tokenize mode)
            token_transformer_list.append(HashTokenTransformer(hashing_secret))
        except Exception as e:
            logger.error("Error initializing hash transformer", exc_info=e)
            raise RuntimeError("Failed to initialize transformer") from e

        try:
            with TokenizeCommand._create_reader(
                input_path, input_type
            ) as reader, TokenizeCommand._create_writer(
                output_path, output_type
            ) as writer:

                metadata = Metadata()
                metadata_map = metadata.initialize()
                # Only record the hashing-secret hash in normal mode
                metadata.add_hashed_secret(Metadata.HASHING_SECRET_HASH, hashing_secret)

                PersonAttributesProcessor.process(
                    reader, writer, token_transformer_list, metadata_map
                )

                MetadataJsonWriter(output_path).write(metadata_map)

        except Exception as e:
            logger.error("Error processing tokens", exc_info=e)
            raise

    @staticmethod
    def _process_tokens_demo(
        input_path: str,
        output_path: str,
        input_type: str,
        output_type: str,
    ):
        """Process tokens in demo mode using PassthroughTokenizer (no hashing)."""
        try:
            with TokenizeCommand._create_reader(
                input_path, input_type
            ) as reader, TokenizeCommand._create_writer(
                output_path, output_type
            ) as writer:

                metadata = Metadata()
                metadata_map = metadata.initialize()
                # Deliberately omit add_hashed_secret — no secret used in demo mode

                PersonAttributesProcessor.process_with_tokenizer(
                    reader, writer, PassthroughTokenizer([]), metadata_map
                )

                MetadataJsonWriter(output_path).write(metadata_map)

        except Exception as e:
            logger.error("Error processing tokens in demo mode", exc_info=e)
            raise

    @staticmethod
    def _create_reader(path: str, file_type: str):
        """Create a PersonAttributesReader based on file type."""
        file_type_lower = file_type.lower()
        if file_type_lower == TokenizeCommand.TYPE_CSV:
            return PersonAttributesCSVReader(path)
        elif file_type_lower == TokenizeCommand.TYPE_PARQUET:
            return PersonAttributesParquetReader(path)
        else:
            raise ValueError(f"Unsupported input type: {file_type}")

    @staticmethod
    def _create_writer(path: str, file_type: str):
        """Create a PersonAttributesWriter based on file type."""
        file_type_lower = file_type.lower()
        if file_type_lower == TokenizeCommand.TYPE_CSV:
            return PersonAttributesCSVWriter(path)
        elif file_type_lower == TokenizeCommand.TYPE_PARQUET:
            return PersonAttributesParquetWriter(path)
        else:
            raise ValueError(f"Unsupported output type: {file_type}")

File: lib/java/opentoken-cli/src/main/java/com/truveta/opentoken/cli/commands/TokenizeCommand.java
/**
 * Copyright (c) Truveta. All rights reserved.
 */
package com.truveta.opentoken.cli.commands;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.concurrent.Callable;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.truveta.opentoken.Metadata;
import com.truveta.opentoken.cli.io.MetadataWriter;
import com.truveta.opentoken.cli.io.PersonAttributesReader;
import com.truveta.opentoken.cli.io.PersonAttributesWriter;
import com.truveta.opentoken.cli.io.RecordIdMappingWriter;
import com.truveta.opentoken.cli.io.csv.PersonAttributesCSVReader;
import com.truveta.opentoken.cli.io.csv.PersonAttributesCSVWriter;
import com.truveta.opentoken.cli.io.json.MetadataJsonWriter;
import com.truveta.opentoken.cli.io.parquet.PersonAttributesParquetReader;
import com.truveta.opentoken.cli.io.parquet.PersonAttributesParquetWriter;
import com.truveta.opentoken.cli.processor.PersonAttributesProcessor;
import com.truveta.opentoken.cli.util.StringMaskingUtil;
import com.truveta.opentoken.tokentransformer.HashTokenTransformer;
import com.truveta.opentoken.tokentransformer.TokenTransformer;
import com.truveta.opentoken.tokens.tokenizer.PassthroughTokenizer;

import picocli.CommandLine.Command;
import picocli.CommandLine.Option;

/**
 * Tokenize command - generates tokens from person attributes.
 *
 * <p>Normal mode (default): applies SHA-256 then HMAC-SHA256 hashing on the token
 * signature, producing opaque base64 tokens. {@code --hashingsecret} is required.
 *
 * <p>Demo mode ({@code --demo-mode}): skips all hashing so tokens are the raw
 * pipe-separated attribute signature strings. No secret is needed, making it easy
 * to explore the output without managing secrets. Demo-mode output is
 * <strong>not</strong> suitable for production or cross-organisation exchange.
 */
@Command(name = "tokenize", description = {
        "Generate tokens from person attributes.",
        "",
        "Normal mode: tokens are HMAC-SHA256 hashed (--hashingsecret required).",
        "Demo mode (--demo-mode): tokens are plain attribute signature strings; no secret needed."
})
public class TokenizeCommand implements Callable<Integer> {

    private static final Logger logger = LoggerFactory.getLogger(TokenizeCommand.class);
    private static final String TYPE_CSV = "csv";
    private static final String TYPE_PARQUET = "parquet";

    @Option(names = { "-i", "--input" }, required = true, description = "Input file path")
    private String inputPath;

    @Option(names = { "-o", "--output" }, required = true, description = "Output file path")
    private String outputPath;

    @Option(names = { "-t", "--input-type" }, required = true, description = "Input file type: csv or parquet")
    private String inputType;

    @Option(names = { "-ot",
            "--output-type" }, description = "Output file type (defaults to input type): csv or parquet")
    private String outputType;

    @Option(names = {
            "--demo-mode" }, description = "Enable demo mode: output raw pipe-separated attribute signature strings with no hashing."
                    + " --hashingsecret is not required in this mode."
                    + " Demo output is NOT suitable for production or cross-organisation exchange.")
    private boolean demoMode;

    @Option(names = { "-h",
            "--hashingsecret" }, description = "Hashing secret for HMAC-SHA256 token generation (required in normal mode)")
    private String hashingSecret;

    @Option(names = { "--help" }, usageHelp = true, description = "Show this help message and exit")
    private boolean helpRequested;

    @Option(names = { "-V", "--version" }, versionHelp = true, description = "Print version information and exit")
    private boolean versionRequested;

    @Option(names = {
            "--hash-record-ids" }, description = "Hash input RecordId values using SHA-256 before writing to output."
                    + " A mapping file (<output>.record-id-mapping.csv) is also written"
                    + " so that hashed IDs can be reconciled back to the originals.")
    private boolean hashRecordIds;

    @Override
    public Integer call() {
        if (demoMode) {
            logger.warn("Running in DEMO MODE - tokens are raw attribute signature strings with no hashing."
                    + " Do not use demo-mode output in production or share it externally.");
        } else {
            logger.info("Running tokenize command (normal mode)");
        }

        // Default output type to input type if not specified
        if (outputType == null || outputType.isEmpty()) {
            outputType = inputType;
        }

        // Log parameters (mask secret only when present)
        logger.info("Input: {} ({})", inputPath, inputType);
        logger.info("Output: {} ({})", outputPath, outputType);
        if (!demoMode && logger.isInfoEnabled()) {
            logger.info("Hashing Secret: {}", maskString(hashingSecret));
        }
        if (hashRecordIds) {
            logger.info("Record ID hashing enabled");
        }

        // Validate file types
        if (!isValidType(inputType)) {
            logger.error("Invalid input type: {}. Must be 'csv' or 'parquet'", inputType);
            return 1;
        }
        if (!isValidType(outputType)) {
            logger.error("Invalid output type: {}. Must be 'csv' or 'parquet'", outputType);
            return 1;
        }

        // --hashingsecret is required in normal mode only
        if (!demoMode && (hashingSecret == null || hashingSecret.isBlank())) {
            logger.error("--hashingsecret is required in normal mode. Use --demo-mode to skip hashing.");
            return 1;
        }

        try {
            processTokens();
            logger.info("Token generation completed successfully");
            return 0;
        } catch (Exception e) {
            logger.error("Error during token generation", e);
            return 1;
        }
    }

    private void processTokens() throws IOException {
        Metadata metadata = new Metadata();
        Map<String, Object> metadataMap = metadata.initialize();

        String mappingFilePath = hashRecordIds
                ? RecordIdMappingWriter.buildMappingFilePath(outputPath)
                : null;

        try (PersonAttributesReader reader = createReader(inputPath, inputType);
                PersonAttributesWriter writer = createWriter(outputPath, outputType);
                RecordIdMappingWriter mappingWriter = hashRecordIds
                        ? new RecordIdMappingWriter(mappingFilePath)
                        : null) {

            if (demoMode) {
                // Skip SHA-256 and HMAC: use PassthroughTokenizer so tokens are the
                // raw pipe-separated attribute signature strings.
                PersonAttributesProcessor.process(reader, writer,
                        new PassthroughTokenizer(List.of()), metadataMap);
            } else {
                // Only record the hashing-secret hash in normal mode
                metadata.addHashedSecret(Metadata.HASHING_SECRET_HASH, hashingSecret);
                PersonAttributesProcessor.process(reader, writer, buildHashTransformers(), metadataMap,
                        mappingWriter);
            }
            MetadataWriter metadataWriter = new MetadataJsonWriter(outputPath);
            metadataWriter.write(metadataMap);
        } catch (Exception e) {
            throw new IOException("Failed to process tokens", e);
        }

        if (hashRecordIds) {
            logger.info("Record ID mapping file written to: {}", mappingFilePath);
        }
    }

    private List<TokenTransformer> buildHashTransformers() throws IOException {
        List<TokenTransformer> transformers = new ArrayList<>();
        try {
            transformers.add(new HashTokenTransformer(hashingSecret));
        } catch (Exception e) {
            throw new IOException("Failed to initialize hash transformer", e);
        }
        return transformers;
    }

    private PersonAttributesReader createReader(String path, String type) throws IOException {
        return switch (type.toLowerCase()) {
            case TYPE_CSV -> new PersonAttributesCSVReader(path);
            case TYPE_PARQUET -> new PersonAttributesParquetReader(path);
            default -> throw new IllegalArgumentException("Unsupported input type: " + type);
        };
    }

    private PersonAttributesWriter createWriter(String path, String type) throws IOException {
        return switch (type.toLowerCase()) {
            case TYPE_CSV -> new PersonAttributesCSVWriter(path);
            case TYPE_PARQUET -> new PersonAttributesParquetWriter(path);
            default -> throw new IllegalArgumentException("Unsupported output type: " + type);
        };
    }

    private boolean isValidType(String type) {
        return TYPE_CSV.equalsIgnoreCase(type) || TYPE_PARQUET.equalsIgnoreCase(type);
    }

    private String maskString(String input) {
        return StringMaskingUtil.maskString(input);
    }
}
